ngraph/core/src/validation_util.cpp

```diff
@@ -1194,3 +1196,375 @@ void ngraph::evaluate_nodes(std::map<RawNodeOutput, HostTensorPtr>& value_map,
         evaluator.evaluate(value);
     }
 }
+
+bool ngraph::could_propagate(const Output<Node>& output, NodeVector& order)
+{
+    bool status = true;
+
+    std::deque<shared_ptr<Node>> nodes_to_calculate = {output.get_node_shared_ptr()};
+    order.push_back(output.get_node_shared_ptr());
+
+    while (status && !nodes_to_calculate.empty())
+    {
+        auto current_node = nodes_to_calculate.front();
+        nodes_to_calculate.pop_front();
+
+        if (current_node->inputs().empty() && !is_type<op::Constant>(current_node))
+            status = false;
+        else if (!is_type<op::v0::ShapeOf>(current_node) && !is_type<op::v3::ShapeOf>(current_node))
+        {
+            // not a leaf, not a shape_of -- continue to search
+            for (const auto& input_value : current_node->input_values())
+            {
+                const auto& input_node = input_value.get_node_shared_ptr();
+                order.push_back(input_node);
```

Comment:
You need also to check that `input_node` is not used yet. Otherwise you can visit the same node more than once.

Reply:
Also I think we definitely need tests for such cases.